pkg/plugins/envvars/envvars.go

```diff
@@ -144,40 +145,57 @@ func (s *Service) GetConfigMap(ctx context.Context, _ string, _ *auth.ExternalSe
 		m[proxy.PluginSecureSocksProxyServerName] = s.cfg.ProxySettings.ServerName
 	}
 
-	// TODO add support via plugin SDK
-	//azureSettings := s.cfg.Azure
-	//if azureSettings != nil {
-	//	if azureSettings.Cloud != "" {
-	//		m[azsettings.AzureCloud] = azureSettings.Cloud
-	//	}
-	//
-	//	if azureSettings.ManagedIdentityEnabled {
-	//		m[azsettings.ManagedIdentityEnabled] = "true"
-	//
-	//		if azureSettings.ManagedIdentityClientId != "" {
-	//			m[azsettings.ManagedIdentityClientID] = azureSettings.ManagedIdentityClientId
-	//		}
-	//	}
-	//
-	//	if azureSettings.UserIdentityEnabled {
-	//		m[azsettings.UserIdentityEnabled] = "true"
-	//
-	//		if azureSettings.UserIdentityTokenEndpoint != nil {
-	//			if azureSettings.UserIdentityTokenEndpoint.TokenUrl != "" {
-	//				m[azsettings.UserIdentityTokenURL] = azureSettings.UserIdentityTokenEndpoint.TokenUrl
-	//			}
-	//			if azureSettings.UserIdentityTokenEndpoint.ClientId != "" {
-	//				m[azsettings.UserIdentityClientID] = azureSettings.UserIdentityTokenEndpoint.ClientId
-	//			}
-	//			if azureSettings.UserIdentityTokenEndpoint.ClientSecret != "" {
-	//				m[azsettings.UserIdentityClientSecret] = azureSettings.UserIdentityTokenEndpoint.ClientSecret
-	//			}
-	//			if azureSettings.UserIdentityTokenEndpoint.UsernameAssertion {
-	//				m[azsettings.UserIdentityAssertion] = "username"
-	//			}
-	//		}
-	//	}
-	//}
+	// Changes here need to be reflected in https://github.com/grafana/grafana-plugin-sdk-go/tree/main/backend/config.go#L152
+	azureSettings := s.cfg.Azure
+	azureDatasources := []string{"grafana-azure-monitor-datasource", "prometheus", "grafana-azure-data-explorer-datasource"}
```

Comment:
right now, this is being forwarded for all plugins AFAIK. Any reason to limit it now to a subset of plugins?

Reply:
Good call out. There were conversations about reducing the size of the request payload. However, I'm uncertain on this and may remove it. I added it here to gather thoughts so we'd have them recorded in comments before making a decision either way.

Reply:
I see, I wouldn't go for the optimization just yet (we could add this later on) but let's see if @wbrowne thinks otherwise

Reply:
Yeah we discussed this as I think it's a good idea long term. Ideally it wouldn't be possible for an unrelated plugin to take advantage of the fact that some other plugin requires Azure config (whether secretive or not). Also, considering the number of config involved for Azure, it would be best to optimize where possible. Best case would be if there was something at the datasource config level where we could decide these cases in general (but I guess isn't the case here).

We don't have to do this now, but I guess why not? 🤔

Reply:
In the non HG world the plugins will still have access to the Azure config via environment so I feel like it should be possible for us to check if any plugins are expecting to use Azure variables that may be affected by this right?

Reply:
>In the non HG world the plugins will still have access to the Azure config via environment

This will be deprecated eventually (ideally the next major Grafana version) - so all plugins should read this config the same way (IE from the request).

> I feel like it should be possible for us to check if any plugins are expecting to use Azure variables that may be affected by this right?

Since this is new/added behaviour, nothing will break until we remove support for env access. If we start with being more restrictive here, then we can see if we need to open this up more if any community plugins also require it. Otherwise, I'm not sure how we'll know?

Reply:
Okay cool, I'm down to keep this in then if that's our consensus. If anything breaks then we're still safe with env access supported.

Reply:
> We don't have to do this now, but I guess why not? 🤔

IMO this is adding a bit of technical debt. This would be yet another list of hardcoded plugin IDs (which doesn't work with things like aliases) and we would need to maintain and potentially add other community/private plugins that would want to use this config.

> In the non HG world the plugins will still have access to the Azure config via environment

I believe this feature is mostly for on-prem? (e.g. AzMG). This is not something that can be easily configured in HG. Since the goal of this config is to have it in Azure-focused environments, I'd assume that most of data sources there would use the config so the optimization would not affect that much of other traffic.

So in general, I'd avoid to add technical debt (even if it's a small amount) for optimization that it's not proven useful :shrug:

Reply:
> IMO this is adding a bit of technical debt. This would be yet another list of hardcoded plugin IDs (which doesn't work with things like aliases) and we would need to maintain and potentially add other community/private plugins that would want to use this config.

On the other hand, should we be sending potentially sensitive information to plugins that probably *shouldn't* see it? Some of these variables can contain secrets and whilst we've done so up until now I don't think that's a good reason to continue the practice.

In HG we don't need to be concerned about private plugins that may need access. I agree there's technical debt to maintaining a hardcoded list but I don't really see the point in sending information to plugins when that information is of no concern to them and they haven't requested it. 

> This is not something that can be easily configured in HG.

For now yes, but that may change. Also, with user-based auth coming at some point we may see more users configuring their instances using these variables.

You raise a good point on if this optimisation is useful, that I don't know..

Reply:
It's only a minor concern so we can also try with the hardcoded list for the moment if you think it's worth it. 

It's true that it's slightly better from a security perspective. In any case, I wouldn't recommend to use config to store secrets since it's not encrypted, but it's already there.

> Maybe it could be a config instead, with some preconfigured default values?

That's a good idea, you are thinking on something like:

```ini
[azure]
forward_settings_to_plugins = grafana-azure-monitor-datasource, prometheus, grafana-azure-data-explorer-datasource
```

Right? That'd be nicer since the list could be changed dynamically but we would need to either change the current env var solution to also not forward it to all plugins (which can be a breaking change?) or not change it (which can be confusing for people reading the setting).

I'll let the decission up to you, I guess we could go ahead with how it's at the moment.

Reply:
>That's a good idea, you are thinking on something like:

Yeah exactly. I think we can just use this for the _new way_. We could document it as such to make it clear for people, or not document at all until the other one is officially deprecated.

Reply:
Okay @andresmgot @wbrowne, I think this is ready for a re-review now.

I've done the following:

Added a property in the `AzureSettings` struct (`ForwardSettingsPlugins`). This is a slice of strings that has a corresponding Grafana config variable `forward_settings_to_plugins`. 

The default behaviour is to have this config variable empty. This will only forward the settings to Grafana owned plugins (Azure Monitor, Azure Data Explorer, Prometheus, MSSQL). A user could add the plugin ID of any plugin they wish and the settings will be forwarded to that plugin(s) in addition to the Grafana owned plugins.

What do we think of this behaviour? This also does not break the current env var solution and I think we're fine to leave the current solution as is to avoid changing the expected behaviour. I've specified in the description of the new variable that this applies to plugin context but I can make it clearer if needed.

Reply:
Sounds good, my preference would be to have the list of plugins in the `defaults.ini` rather than in the code. It's more flexible and also self-document how people can use it (e.g. using plugin IDs, not names or something like that).

Reply:
That's a reasonable change, I was torn between that and having them defined in code so they will always be set for GL owned plugins but I think being flexible is probably better 😊